### ISSUE TITLE
`no-undef` reports even though variable exists.

### ISSUE THREAD

**Issue by Yash-Singh1** [eslint/eslint#15026]
**Environment**

* **ESLint Version:** ^7.32.0
* **Node Version:** v16.3.0
* **npm Version:** 7.19.1
* **Operating System:** Ubuntu 20.04.3 LTS

**Parser**

default

**Configuration**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  rules: {
    'no-undef': 'error'
  }
}
```

</details>

**Code**

```js
/* eslint-env browser */

if (typeof stuff !== 'function') {
  window.stuff = function name(params) {};
}

stuff();
```

**Command**

```bash
eslint ./**/*.js
```

**What did you expect to happen?**

No errors

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

```console
7:1 - 'stuff' is not defined. (no-undef)
```

**Steps to reproduce this issue:**

1. Open [playground](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50LWVudiBicm93c2VyICovXG5cbmlmICh0eXBlb2Ygc3R1ZmYgIT09ICdmdW5jdGlvbicpIHtcbiAgd2luZG93LnN0dWZmID0gZnVuY3Rpb24gbmFtZShwYXJhbXMpIHt9O1xufVxuXG5zdHVmZigpO1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby11bmRlZiI6Mn0sImVudiI6e319fQ==)
2. Notice how there is an error even though we are sure that `stuff` exists

**Are you willing to submit a pull request to fix this bug?**

I am unsure how to fix this issue.

**Comment by snitin315** [comment]
In the given [playground](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50LWVudiBicm93c2VyICovXG5cbmlmICh0eXBlb2Ygc3R1ZmYgIT09ICdmdW5jdGlvbicpIHtcbiAgd2luZG93LnN0dWZmID0gZnVuY3Rpb24gbmFtZShwYXJhbXMpIHt9O1xufVxuXG5zdHVmZigpO1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby11bmRlZiI6Mn0sImVudiI6e319fQ==) , there is no declaration for variable `stuff`. Hence, ESLint will consider it undefined unless it is a global variable.

See [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50LWVudiBicm93c2VyICovXG5cbmNvbnN0IHN0dWZmID0gXCJzdHVmZlwiO1xuXG5pZiAodHlwZW9mIHN0dWZmICE9PSAnZnVuY3Rpb24nKSB7XG4gIHdpbmRvdy5zdHVmZiA9IGZ1bmN0aW9uIG5hbWUocGFyYW1zKSB7fTtcbn1cblxuc3R1ZmYoKTtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsibm8tdW5kZWYiOjJ9LCJlbnYiOnt9fX0=) for example.

In case `stuff` is a global variable you can add a comment at the top `/*global stuff*/`  and ESLint will consider it as defined.

```js
/* eslint-env browser */
/* global stuff */

if (typeof stuff !== 'function') {
  window.stuff = function name(params) {};
}

stuff();
```
Checkout [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50LWVudiBicm93c2VyICovXG4vKiBnbG9iYWwgc3R1ZmYgKi9cblxuaWYgKHR5cGVvZiBzdHVmZiAhPT0gJ2Z1bmN0aW9uJykge1xuICB3aW5kb3cuc3R1ZmYgPSBmdW5jdGlvbiBuYW1lKHBhcmFtcykge307XG59XG5cbnN0dWZmKCk7XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXVuZGVmIjoyfSwiZW52Ijp7fX19), it's lint free.

**Comment by Yash-Singh1** [comment]
Let's assume that this is wrapped inside a function that will be called at some point of time in the application. Then, it shouldn't be considered global, because the `stuff` variable doesn't exist yet, until that function is called. Example:

```js
/* eslint-env browser */
/* global stuff */

function A() {
  if (typeof stuff !== "function") {
    window.stuff = function name(params) {};
  }

  stuff();
}
stuff; // <--- should be error
A();
stuff; // <--- Good to go
```

**Comment by snitin315** [comment]
I see, the problem here is that `stuff` is an unknown global variable. This issue was brought up in https://github.com/eslint/eslint/issues/7636 and https://github.com/eslint/eslint/issues/6209 before but was closed as consensus couldn't be reached.

Anyways this can also work:
 
```js
/* eslint-env browser */

if (typeof stuff !== 'function') {
  window.stuff = function name(params) {};
}

window.stuff();
```

Let's wait for feedback from other team members if they think this change is worth it to be accepted.
